tests/integrationv2/test_renegotiate.py

```diff
@@ -0,0 +1,259 @@
+import copy
+import os
+import pytest
+import time
+import random
+from collections import namedtuple
+
+from configuration import available_ports, ALL_TEST_CIPHERS, ALL_TEST_CURVES, MINIMAL_TEST_CERTS, PROTOCOLS
+from common import ProviderOptions, Protocols, Curves
+from fixtures import managed_process
+from providers import Provider, S2N, OpenSSL, GnuTLS
+from utils import invalid_test_parameters, get_parameter_name
+
+
+# TLS1.3 does not support renegotiation
+TEST_PROTCOLS = [x for x in PROTOCOLS if x.value < Protocols.TLS13.value]
+
+
+# Command line options to enable renegotiation
+S2N_RENEG_OPTION = "--renegotiation"
+S2N_RENEG_ACCEPT = "accept"
+S2N_RENEG_REJECT = "reject"
+
+
+# Output indicating renegotiation state
+S2N_RENEG_START_MARKER = "RENEGOTIATE"
+S2N_RENEG_SUCCESS_MARKER = "s2n is ready, again"
+OPENSSL_RENEG_REQ_MARKER = "SSL_accept:SSLv3/TLS write hello request"
+OPENSSL_RENEG_WARN_MARKER = "SSL3 alert read:warning:no renegotiation"
+
+
+# Methods to check renegotiation state
+def renegotiate_was_requested(openssl_results):
+    return to_bytes(OPENSSL_RENEG_REQ_MARKER) in openssl_results.stderr
+def renegotiate_was_rejected(openssl_results):
+    return to_bytes(OPENSSL_RENEG_WARN_MARKER) in openssl_results.stderr
+def renegotiate_was_started(s2n_results):
+    return to_bytes(S2N_RENEG_START_MARKER) in s2n_results.stdout
+def renegotiate_was_successful(s2n_results):
+    return renegotiate_was_started(s2n_results) and \
+           to_bytes(S2N_RENEG_SUCCESS_MARKER) in s2n_results.stdout
+
+
+# Basic conversion methods
+def to_bytes(val):
+    return str(val).encode('utf-8')
+def to_marker(val):
+    return bytes(val).decode('utf-8')
+
+
+"""
+The integrationv2 framework controls process IO with "markers".
+Messages can only be written to stdin if "send_marker"s are read from stdio.
+This results in the restriction that a process can only send in response to its own output,
+and also makes reasoning about message ordering very difficult.
+
+Because testing renegotiation requires multiple messages to be sent and received,
+we construct the "data_to_send", "send_marker" and "close_marker" lists required
+by the integrationv2 framework programmatically based on a "Msg" abstraction
+that is easier to visualize and reason about.
+
+For simplicity, "Msg" currently assumes:
+- s2n sends first. This lets us assume the first send_marker is S2N.get_send_marker().
+- The server sends last. This lets us assume that only the client needs a close_marker.
+"""
+class Msg(object):
+    def __init__(self, mode, send_marker=None, ctrl_str=None):
+        self.mode = mode
+        # Indicates what stdio string should trigger this message.
+        # Will default to the previous message (see Msg.send_markers).
+        self.send_marker = send_marker
+        # Indicates that the message will not be received by the peer
```

Comment:
This comment is kind of obtuse. ctrl_str is used to trigger behavior; the fact that it isn't received by the peer is kind of irrelevant.

Reply:
It is relevant to IO. It means that it will never appear in the peer's stdio, which is a potential complication for an IO system where every step needs an stdio trigger. This class is all about managing the framework's weird IO requirements.

But let me see if I can rephrase to cover both pieces of information.